Refactor and rework documentation on astropy file I/O

taldcroft opened the pull request:
<!-- These comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
https://docs.astropy.org/en/latest/development/quickstart.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/git_edit_workflow_examples.html . -->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->

The current astropy narrative documentation on I/O is quite disorganized and confusing to follow. There is no clear overview of the components and a good deal of repetition.

This PR is an attempt to rectify this situation:
- Introduce explicit labels "High-level Unified I/O" and "Low-level I/O".
- Add a new Overview section that describes the distinction between high- and low-level I/O.
- Break the Table Unified I/O into sub-documents. The current single document is way too long.
- General (though not comprehensive) migration from "ASCII" tables to "text" tables, since unicode text is fully supported.
- Move SAMP and the `io.registry.typing` sections out of I/O, so that the I/O section is consistently about reading and writing data files.
- Break the text format tables into categories.
- Remove some mostly-redundant content, e.g. `docs/table/io.rst`.

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

<!-- Optional opt-out -->
Please DO use the "Squash and Merge" button!

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

pllim commented:
Missing import?

```
_________________________ [doctest] unified_table.rst _________________________
256 
257 Then, we can read back that single-column table and extract the column from it::
258 
259     >>> col = Table.read('obstime.fits').columns[0]
260     >>> type(col)
261     <class 'astropy.table.column.Column'>
262 
263 .. testcleanup::
264 
265     >>> pathlib.Path.unlink('obstime.fits')
UNEXPECTED EXCEPTION: NameError("name 'pathlib' is not defined")
Traceback (most recent call last):
  File "C:\hostedtoolcache\windows\Python\3.12.9\x64\Lib\doctest.py", line 1368, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest unified_table.rst[20]>", line 1, in <module>
NameError: name 'pathlib' is not defined. Did you forget to import 'pathlib'
D:\a\astropy\astropy\docs\io\unified_table.rst:265: UnexpectedException
============================ slowest 50 durations ============================
```

pllim commented:
RTD has a warning (occurs twice): `docs/io/unified_table.rst:98: WARNING: Failed to create a cross reference. A title or caption not found: 'table_io_parquet'`

Don't think the remote failures are related.

taldcroft commented:
Passing tests and RTD, finally!

hamogu commented on docs/io/unified_table_fits.rst:
```diff
@@ -0,0 +1,625 @@
+FITS table HDU, but also any ``astropy`` native objects
+(:ref:`mixin_columns`) within a `~astropy.table.Table` or
+`~astropy.table.QTable`. This includes `~astropy.time.Time`,
+`~astropy.units.Quantity`, `~astropy.coordinates.SkyCoord`, and many others.
```
"many others"? Really? I can't think of "any other". Maybe just say "others that follow the protocol described in the next section"?

hamogu commented on docs/io/unified_table_fits.rst:
```diff
@@ -0,0 +1,625 @@
+represented in a FITS table using the ``TUNITn`` FITS column keyword to
+incorporate the unit attribute of Quantity. For example:
+
+.. doctest-skip::
```
I'll stop commenting on that, but all those doctests should be run and not be skipped.

taldcroft replied:
I was initially annoyed :smile:, but it seems to be going well and fixing a few mistakes along the way.

hamogu commented on docs/io/unified_table_fits.rst:
```diff
@@ -0,0 +1,625 @@
+these should be used by passing ``astropy_native=True`` (for backwards
+compatibility, this is not done by default). This will convert columns
+conforming to the FITS time standard to `~astropy.time.Time` instances,
+avoiding any loss of precision.
```
```suggestion
avoiding any loss of precision and preserving information about the time system if set in the fits header.
```

taldcroft replied:
Done.

hamogu commented on docs/io/unified_table_fits.rst:
I first read this as "arbitrary values", but what you mean is "arbitrary datatypes". If my meta has strings, they will turn up as FITS header cards. Not sure how to say that here to preserve the meaning. "Note the metadata in ... that is not just a simple string is not written an will be lost"?

taldcroft replied:
That sentence about Table meta representation in FITS was completely out of place here, so I just removed it. Sometimes less is more.

hamogu commented on docs/io/unified_table_hdf5.rst:
```diff
@@ -0,0 +1,96 @@
+easily access and use the table metadata if reading the HDF5 in a non-astropy
+application.
+
+As of ``astropy`` 3.0, by specifying ``serialize_meta=True`` one can also store
```
Astropy 3.0 was a while ago. Maybe "You can also store table that contain ... by specifying ``serialize_meta=True``"?

taldcroft replied:
Done.

hamogu reviewed (changes requested):
My main suggestion is to use the opportunity of re-ding this to activate more of the doctests. A lot of doctests are skipped that don't need to be, and running them gives us confidence that the docs stay up to date even in any future changes.

I thin the text in the individual files is good and the structure will help. There is no ideal way that will make everyone happy, but this is way better then what we have now.

taldcroft commented on docs/io/unified_table_fits.rst:
```diff
@@ -0,0 +1,625 @@
+to writing.
+
+At this time, the ``meta`` attribute of the :class:`~astropy.table.Table` class
+is an ordered dictionary and does not fully represent the structure of a
```
Done.

taldcroft commented on docs/io/unified_table_fits.rst:
```diff
@@ -0,0 +1,625 @@
+merged back into one masked column.
+
+.. doctest-skip::
+
```
Fixed.

taldcroft commented on docs/io/unified_table.rst:
```diff
@@ -0,0 +1,294 @@
+
+..
+  EXAMPLE END
+
```
Done.

taldcroft commented on docs/io/unified_table_fits.rst:
```diff
@@ -0,0 +1,625 @@
+keyword::
+
+    >>> t.write('existing_table.fits', append=True)  # doctest: +SKIP
+
```
Done.

taldcroft commented on docs/io/unified_table_fits.rst:
```diff
@@ -0,0 +1,625 @@
+- The masked data values are permanently lost, precluding the possibility
+  of later unmasking the values.
+
+``astropy`` provides a work-around for this limitation that users can choose to
+use. The key part is to use the ``serialize_method='data_mask'`` keyword
+argument when writing the table. This tells the FITS writer to split each masked
```
I agree with your 90% of users use 10% of astropy. Including me.

hamogu commented on docs/io/unified_table_fits.rst:
```diff
@@ -0,0 +1,632 @@
+
+To read a FITS table into `~astropy.table.Table`:
+
+.. doctest-skip::
```
Why is this dockets skipped?

hamogu commented on docs/io/unified_table_fits.rst:
Why skip this doctest?

hamogu commented on docs/io/unified_table_fits.rst:
Why skip this doctest?

hamogu reviewed (changes requested):
Sorry, two more doctests... I didn't mark them all on initial review, because I wasn't sure if there is a good reason that I didn't see why we don't run them.

hamogu commented on docs/io/unified_image.rst:
```diff
@@ -0,0 +1,35 @@
+
+.. doctest-skip::
+
+    >>> # Read CCD image
+    >>> ccd = CCDData.read('image.fits')
```
```suggestion
    >>> # Read CCD image
    >>> from astropy.nddata import CCDData
    >>> ccd = CCDData.read('image.fits')
```

hamogu commented on docs/io/unified_table.rst:
```diff
@@ -0,0 +1,293 @@
+are expanded in the same way as is done by many command-line utilities, to
+represent the home directory of the current or specified user, respectively.
+
+..
+
+  EXAMPLE END
```
Need to remove empty line or sphinx. will render the text "EXAMPLE END" in a box.

```suggestion
..
  EXAMPLE END
```

hamogu commented on docs/io/unified_table.rst:
```diff
@@ -0,0 +1,293 @@
+The two classes which have configurable serialization methods are `~astropy.time.Time`
+and `~astropy.table.MaskedColumn`. See the sections on Time
+:ref:`unified_table_fits_details` and :ref:`unified_table_fits_masked_columns`,
+respectively, for additional information. The defaults for each format are listed below:
```
"for additional information how this is handled in fits files"? The links leads to the fits section, while from the text here I expected it to be general for all formats. Instead, som general info is in the next few sentences, but I don't think we have (and maybe also don't need) a section with more details for all formats.

saimn commented:
I like the general idea but there is a drawback with moving the io.fits, io.ascii etc. docs one level down, we loose the left sidebar navigation for the subsections. 

![image](https://github.com/user-attachments/assets/f6bf2b74-b5da-41c6-ab8a-017b2373bc9d)

vs 

![image](https://github.com/user-attachments/assets/7ede5b39-8a5e-4c5e-a67c-28aa02ad52fd)

Can we just keep them at the same level as high and low level pages ?

taldcroft commented:
> Can we just keep them at the same level as high and low level pages ?

I see your point about the sections in the FITS page. 

The issue is that this breaks the clean division between high and low level I/O. Part of the goal is to strongly guide users to read the docs and get an understanding of the distinction. Then there is also the question of what to expose at the top level. VOtable? The `misc` stuff like YAML?

I think that a fair number of users land on the current doc pages with a goal of reading a FITS table file. Being impatient they see the word `FITS` at the top level and click on that. The next big text that catches their eye is the section header `Reading and Updating Existing FITS Files` and they are off and running using `fits.open()`. Same story for `ascii.read()`.

It may be controversial or too much, but the idea is to purposely require another navigation step to get there. So I'm not quite sure how to thread the needle.

nstarman commented:
It's true. I know so many people that always use the low-level fits opener rather than `QTable.read`.
Making the low-level functions *less* apparent is actually a +1 for me.

saimn commented:
There are many things that cannot be done with the high level api. So I think it's good to highlight the high level api but it doesn't have to be at the expense of the low level one. Maybe make 2 sections ?

taldcroft commented:
@saimn - understood. I have a compromise solution in mind that I hope will work and address your concerns. Hopefully I'll get this done along with addressing all the new comments from @hamogu.

hamogu reviewed (approved):
My concerns are all addressed, I know that others wanted to discuss more on how prominent the low level interface should be, so I'm not merging yet, but I'm approving from my perspective.

taldcroft commented:
@hamogu et al. - there was a merge conflict with #17828 that (with a bit of difficulty) I have resolved. In doing so, I squashed all the commits that have been previously reviewed. Now there are four more separate commits that are worth reviewing individually.

I hope the outcome is acceptable to all. The low-level I/O packages are back to the original level. The "compromise" is that for FITS, io.ascii and VOTable, I moved the note about using the high-level interface to a more prominent location.

One commit is to replace "ASCII" with "text" in a lot of places, since `io.ascii` actually reads unicode text (except for the fast reader).

taldcroft commented:
@hamogu - hoping for just one more review. :smile:

hamogu commented:
I understand that sometimes people may need `io.fits` (I sometimes do) and thus we don't worry it too deep. But with this, if I'm new to the docs, and I need to read a fits file (which is probably the most common use case why someone installs astropy), I will click on the heading that says "fits" and I'm immediately in the low-level interface. Same if I want to read a text or vo table. 
What does "high level unified IO" mean anyway? That's it's a highly complicated way to read files and needs a high-level of programming skills? Maybe we can change the section heading for that to "Read any file (fits, text, vo, and more)" so that that word "fits" appears in there as well? Though "read any file" still sounds like "General intro to file reading and then you need to click on FITS file handling anyway, so why don't you save yourself some time and go to the section on fits file handling immediately".

Can we do better? Maybe "One method to rule them all: Fits, text, vo and others all in one fell swoop". But that might be a little too long and too informal. Mmm. At some point we paid a technical writer and now when we need one we have a 4 year old handbook telling me if I should say "we" or "you" or "the user" in the docs, but no advice on how to deal with this. So, I'll continue to try myself:

"The simple way to read/write any file (fits, text, vo, and others)"

Not ideal, I get that, but the best I can come up with. More suggestions welcome! But it has all the requirements I developed above:
- It avoids technical language such as "high-level" and "unified IO".
- It contains the word "fits" to entice the new user to actually click on this section heading (and to increase the chance the google served this page when I google "astropy read fits", which currently (and understandably) gets me to the `io.fits` docs (second is a learn astropy tutorial that also uses `io.fits` instead of the high-level unified IO interface and then come several hits for subpages within `io.fits`). Without the word "fits" in the heading and the links to the high-level pages, I don't think google will offer that page in any way. Even if I google "astropy high-level IO interface", it gives me the `io.fits` pages because that has text that says "CCDData provides a higher level interface to read and write FITS files ..." and the `io.fits` pages is probably very high in it's page rank because so many people from all over the world link to it.

saimn commented on docs/io/fits/index.rst:
```diff
@@ -11,8 +11,13 @@ Introduction
 The :mod:`astropy.io.fits` package provides access to FITS files. FITS
 (Flexible Image Transport System) is a portable file standard widely used in
-the astronomy community to store images and tables. This subpackage was
-originally developed as PyFITS.
```
Not sure if this PR is the good place to remove this sentence. PyFITS is mentioned a few times in the docs, including a few lines below, and I think it's good to state clearly that `io.fits` comes from PyFITS.

taldcroft replied:
Just curious about why. That fork was like 14 years ago, so I think that this history should not be the 3rd piece of information that new users learn about `astropy.io.fits`. If there needs to a nod to the history I would put that in an acknowledgements section at the end. PyFITS itself has been deprecated for 8 years and is unlikely to work with modern Python.

saimn replied:
PyFITS is mentioned in the docs (in this page) so this sentence make it clear what PyFITS is (for people that don't know it) or what it became. It might be time to revisit how we mention it (and link to the [history page](https://docs.astropy.org/en/latest/io/fits/appendix/history.html)) but that deserves its own PR imo.

taldcroft replied:
OK, I'll revert for this PR. But there really should be a large followup PR to reflect the fact that PyFITS is now strictly just historical legacy. Searching for PyFITS in the RST docs shows LOTS of just-plain-wrong statements, like: 

The `astropy.io.fits` module is identical to PyFITS but with the names changed.
... There are plans to gradually phase out PyFITS as a stand-alone
module and deprecate it in favor of `astropy.io.fits`.

Basically there is a lot of historical cruft that can only serve to confuse or distract current users.

mhvk commented on docs/io/unified_table_fits.rst:
```diff
@@ -0,0 +1,629 @@
+.. _fits_astropy_native:
+
+
+TDISPn Keyword
```
My tendency would be to put this nearer the end: I think this is a detail that matters, but less so on how time columns, etc., are dealt with.

taldcroft replied:
I'm not sure, but I think this is too far out of scope. It is an an existing detail of the Table-FITS documentation that has not changed in this PR, not about the big-picture refactor.

saimn commented on docs/io/unified_table_fits.rst:
```diff
@@ -0,0 +1,629 @@
+
+You can also read a table from the HDUs of an in-memory FITS file. This will
+round-trip any :ref:`mixin_columns` that were written to that HDU, using the
+header information to reconstruct them::
```
But round trip is not specific to the in-memory file case ?

taldcroft replied:
Fixed.

mhvk reviewed (commented):
@taldcroft - now finally found some time to look. Overall, very much an improvement. Some smaller organizational comments in-line, mostly for the landing page, which I think is most important to get right.

Also a more general question: my tendency would be to lead users to `QTable` rather than `Table`, as units in general are not meaningless labels. So, for `unified_table` in particular, maybe just use `QTable` throughout?

Another general comment echoes @hamogu's - it would be really nice if the examples in the documentation were not skipped - for writing, this is in principle not difficult.

Of course, I can totally see the argument that the above two larger changes could/should be decoupled from the re-organization that you have done...

taldcroft commented on docs/index_user_docs.rst:
```diff
@@ -20,16 +20,19 @@ User Guide
+   :caption: File I/O
+   :maxdepth: 2
 
+   io/overview
```
I'd rather keep it the way it is.
- Having short and clearly delineated topic pages is a feature, not a bug.
- By combining like you suggested, then the explanation of Unified I/O is buried in the Overview and thus lost to impatient readers.
- One of the prime drivers of this entire PR was to make a clear, short, and highly visible explanation of the basic concept. This is unique to astropy and so it needs to be highlighted.
- Using the 2nd level hierarchy for Image Data and Table Data makes it clear that they are part of the unified interface without the somewhat awkward repetition of `via unified I/O`.

mhvk replied:
Given @hamogu's points how about keeping your structure but making the headings more specific,
```
Standard file I/O
    FITS images
    Text, FITS, and other tables
```

One could also think of changing the title of the following ones to `Low-level FITS interface` or `Expert FITS I/O (astropy.io.fits)` or something like that.

Though perhaps what you had before, a 2-level also for the low-level interfaces, is better...

taldcroft commented:
@hamogu - Like you, I've been struggling with this. The docs are caught between different needs and this PR is trying to balance those out as much as possible. There are some ground rules that tightly constrain possible solutions:

- This PR is intended as a refactor and re-organization of existing concepts and labels. I've taken some liberties like `ASCII` -> `text` and trying to tighten content, but mostly the goal is moving stuff and providing the necessary conceptual scaffolding.
- "Unified I/O" has been around for a very long time. Changing that terminology is something Astropy can consider but it should be out of scope for this PR .
- In order to maintain the current left-bar navigation in `io.fits` and `io.ascii`, those sections need to stay at the top level.
- `io.fits` is really a hybrid. For reading a table, in most (but not all) cases using the unified interface is preferred, but for reading a FITS image to a numpy array, I would use `io.fits` instead of `CCDData.read()` (since I'm not reading CCD images).
  - In other words, I agree that keeping FITS at the top level is the right decision. Hiding it one layer down is not a good idea.

The solution in this PR to users clicking on `FITS` or `io.ascii` or finding it through google is the more prominent note RIGHT ON TOP alerting users that they might want to use the unified I/O reader.

pllim commented:
I see that I was pinged to review but `+1,512 −1,436` and 100 comments later, perhaps my review is not necessary here. Would be happy to chime in if there are specifics you want me to focus on. Otherwise, I trust you all as long as CI is green. 😸

taldcroft commented on docs/io/unified_table_fits.rst:
```diff
@@ -0,0 +1,629 @@
+The keyword names (e.g., ``MY_KEYWD``) will be automatically capitalized prior
+to writing.
+
+The ``meta`` attribute of the :class:`~astropy.table.Table` class
```
All parts of this sentence look to be entirely redundant with the content above. I'm just taking it out.

taldcroft commented on docs/io/unified_table_fits.rst:
```diff
@@ -0,0 +1,629 @@
+
+.. _unified_table_fits_masked_columns:
+
+Masked Columns
```
Same, out of scope.

taldcroft commented on docs/io/unified_table_hdf5.rst:
```diff
@@ -0,0 +1,96 @@
+To read a table called ``data`` from an HDF5 file named ``observations.hdf5``,
+you can do::
+
+    >>> t = Table.read('observations.hdf5', path='data')
```
Done.

taldcroft commented on docs/io/unified_table_hdf5.rst:
```diff
@@ -0,0 +1,96 @@
+..
+  EXAMPLE END
+
+.. doctest-skip-all
```
Fixed.

taldcroft commented on docs/io/unified_table_hdf5.rst:
```diff
@@ -0,0 +1,96 @@
+
+To write a table to a new file, the path should also be specified::
+
+    >>> t.write('new_file.hdf5', path='updated_data')
```
Sure, but again this is out of scope for the intent of this PR.

taldcroft commented on docs/io/unified_table_text.rst:
```diff
@@ -0,0 +1,186 @@
+  >>> t2 = Table.read('data.csv', format='pandas.csv', sep=' ', names=['a', 'b', 'c'])
+
+
+Supported Formats
```
Done.

taldcroft commented on docs/index_user_docs.rst:
```diff
@@ -40,13 +43,15 @@ User Guide
 
    config/index
    io/registry
+   io/typing
```
For the sake of moving things along, I just did this. If there are objections I can revert.

taldcroft commented:
@mhvk @saimn @hamogu @nstarman - I think I have either resolve or responded to all the comments at this point.

I think this PR has gotten to the point where I hope it can be merged. If there are other small improvements they can be handled with new PR(s). Already this PR is causing a block on #16780 because we cannot update narrative docs there without causing a conflict mess. I'm also holding on #17706 for the same reason.

hamogu reviewed (approved):
I have reviewed (and approved) this in the past. There are some things that I personally would like to be different, but I agree that it's either not important (we'll never all agree on every comma) or not in the scope if this PR.

pllim commented:
Merged. Thanks, all!

pllim commented:
p.s. Oh no, I did not use Squash and Merge and did not see the comment that I should until too late. I thought 7 commits for such big diff was reasonable. Sorry!

mhvk commented:
Thanks, @taldcroft, a very nice improvement!
